tensorboard/plugins/graph/README.md

```diff
@@ -121,9 +121,384 @@ describing the motivations, research, and process behind it.
+            suffix changes. For example, 3 OpNodes with op 'Add': `add_1`,
+            `add_2`, `add_3` will be grouped into a SeriesNode 'add_[1-3]'.
+        -   Series groups require at least 2 nodes (later we filter out groups
+            with less than 5 nodes) where the numeric suffix is in increasing
+            order.
+        -   Children within a SeriesNode have a reference: `child.owningSeries =
+            seriesName`.
+    -   If series has 5+ members,
+        -   Add it to hierarchy and the metagraph, updating references,
+            histograms, etc. The metagraph no longer considers the original
+            OpNodes as direct children, if they are owned by a SeriesNode.
+-   'Adding edges'
+    -   For each BaseEdge
+        -   Instead of creating an edge between the two nodes, create a MetaEdge
+            between their ancestors underneath their common ancestor furthest
+            from the root. Update the shared ancestor MetaNode's metagraph.
+        -   For each metanode with a metaedge, store whether it has a 'non
+            control dependency' edge.
+        -   Compute the size of the tensor flowing through this base edge and
+            store the maxMetaEdgeSize on the Hierarchy itself.
+            -   For each outputShape tensor, get the "size": [4, 2, 1] is 4 *
+                2 * 1 = 12.
+            -   When a metaedge contains multiple base edges, it sums their
+                total sizes
+        -   Update the MetaEdge with a reference to the BaseEdge:
+            `metaedge.addBaseEdge(baseEdge, h)`. A metanode may contain multiple
+            BaseEdges if there are multiple connections between two metanodes.
+    -   Note, the Hierarchy does not have a direct reference to metaedges. It
+        simply has a root metanode with a metagraph, and edges of the metagraph
+        are metaedges.
+-   'Finding similar subgraphs'
+
+    -   let nnGroups = clusterSimilarSubgraphs(h);
+        -   Compute signatures for **all nodes, even non-rendered** ones.
+            Signatures over all nodes in the entire graph are required to
+            determine the templates for node coloring.
+        -   Signatures come from serializing a node's structural info:
+            -   "depth=1 |V|=3 |E|=0 [ops] Add=3"
+            -   "depth=2 |V|=6 |E|=7 [ops] Placeholder=2,Add=9,Pow=1"
+        -   Create a map from `Record<Signature, MetaNode[]>`, and keep only the
+            clusters that contain multiple MetaNodes, unless it is a library
+            function. For functions, we may "add more nodes with the template
+            later" (TODO: why?).
+    -   let templates = groupTemplateAndAssignId(nnGroups, verifyTemplate);
+    -   For each MetaNode in a nnGroup
+
+        -   For all previous clusters, try to add the current MetaNode to the
+            previous cluster if their metagraphs are similar. Note: this likely
+            results in a quadratic+ time complexity.
```

Comment:
Right, actual guestimation of time complexity would be nice in terms of |cluster| and complexity of clustering with |N| and |E|. 

Do you think there is some chance of saving compute time if we went bottom to up instead to achieve memoization?